File: Calculator.py
# Course: cmps3500
# CLASS Project
# PYTHON IMPLEMENTATION OF A CUSTOM MATRIX CALCULATOR
# date: 05/14/2021
# Student 1: Arielle Battle
# Student 2: Janneth Guarcas Garcia
# Student 3: Javier Medina
# Description: Implementation of a Matrix calculator in Python 2

import csv
import numpy as np

menu = {}
menu['\n01']="ADD A and B: A + B"
menu['02']="Subtract A from B: A - B"
menu['03']="Subtract B from A: B - A"
menu['04']="Multiply A and B: AB"
menu['05']="Multiply B and A: BA"
menu['06']="Copy A into B: A to B"
menu['07']="Copy B into A: B to A"
menu['08']="Swap A and B: A to B"
menu['09']="Load matrix: Load A"
menu['10']="Load matrix: Load B"
menu['11']="Make A and square identify matrix up to order 10x10: A to I"
menu['12']="Make B and square identify matrix up to order 10x10: B to I"
menu['13']="Scalar n times A: nAwhere n is an integer"
menu['14']="Scalar n times B: nA where n is an integer"
menu['15']="Determine of A: det(A)"
menu['16']="Determine of B: det(B)"
menu['17']="A transpose: A^T"
menu['18']="A transpose: B^T"
menu['19']="Inverse of A: A^-1"
menu['20']="Inverse of B: B^-1"
menu['21']="Integer power of A when A is a square matrix: A^n for 1 <= n <= 10"
menu['22']="Integer power of B when B is a square matrix: B^n for 1 <= n <= 10"
menu['23']="Exit"

#file for opening a file for A.csv
Afile = open("A.csv")
Afile_numpy = np.gefromtxt(Afile, delimeter = ",", skip_header = 1)

#file for opening a file for B.csv
Bfile = open("A.csv")
Bfile_numpy = np.gefromtxt(Afile, delimeter = ",", skip_header = 1)

#function for add
 
#function for subtract A - B
#function for subtract B - A
#function for multiply AB
#function for multiply BA
#function for copying A into B
#function for copying B into A
#function for making A and square identify matrix up to oder 10x10: A to I
#function for making B and square identify matrix up to oder 10x10: A to I
#function for scalar n times A
#function for scalar n times B
#function for det(A)
#function for det(B)
#function for transpose A
#function for transpose B
#function for inverse A
#function for inverse B
#function for Integer power of A when A is a square matrix
#function for Integer power of B when B is a square matrix

#while its not exit run menu program
while True:

    #get users input
    options=menu.keys()
    options.sort()

    for entry in options:
        print entry, menu[entry]

    #has to be 'raw_input' in order to make selections for menu
    selection = raw_input("\nEnter Choice #: ")
    if selection =='1':
        result = np.sum(Afile_numpy, Bfile_numpy)
        print result

    elif selection == '2':
        result = np.subtract(Afile_numpy, Bfile_numpy)
        print result

    elif selection == '3':
        result = np.subtract(Bfile_numpy, Afile_numpy)
        print result

    elif selection == '4':
        result = np.dot(Afile_numpy, Bfile_numpy)
        print result

    elif selection == '5':
        result = np.dot(Bfile_numpy, Afile_numpy)
        print result

    elif selection == '6':

    elif selection == '7':

    elif selection == '8':

    elif selection == '9':
        #this will open the file and read in the numbers
        with open('A.csv', 'rb') as f:
            reader = csv.reader(f)
            data_as_list = list(reader)
            print data_as_list

        print Afile_numpy

    elif selection == '10':
        #this will open the file and read in the numbers
        B = int (input("input integer for B: "))
        with open('B.csv', 'rb') as f:
            reader = csv.reader(f)
            data_as_list = list(reader)
            print data_as_list

        print Bfile_numpy

    elif selection == '11':

    elif selection == '12':

    elif selection == '13':

    elif selection == '14':

    elif selection == '15':

    elif selection == '16':

    elif selection == '17':

    elif selection == '18':

    elif selection == '19':

    elif selection == '20':

    elif selection == '21':

    elif selection == '22':

    elif selection == '23':
        print("\nExiting program")
        break

    else:
        print("\nMust be a number on menu!\n")

File: Calculator.rb
#!/usr/bin/env ruby
# #############################################################################
# Course: cmps3500
# CLASS Project
# Ruby IMPLEMENTATION OF A CUSTOM MATRIX CALCULATOR
# date: 05/14/2021
# Student 1: Arielle Battle
# Student 2: Janneth Guarcas Garcia
# Student 3: Javier I. Medina
# Description: Implementation of a Matrix calculator in Ruby 05082021
# #############################################################################

require 'csv'

# #############################################################################
# function definitions 

# unary operations per requirements

# NOTES: 

def load_A
    matrix = CSV.parse(File.read("A.csv"), converters: :integer)
end

def load_B
    matrix = CSV.parse(File.read("B.csv"), converters: :integer)
end

# MENU 3*** make A and square indentity matrix up to oder 10x10: A to I ***
def A_to_I
    #code here
end

# MENU 4*** make B and square indentity matrix up to oder 10x10: A to I ***
def B_to_I
    #code here
end

def scale_A(n, matrix_A)
    row = matrix_A[0].length
    col = matrix_A.length
    new_A = Array.new(col){Array.new(row)}

    puts "Scaling Matrix A by n = #{n}"

    i = 0
    matrix_A.each { |array|
        j = 0
        puts array
        array.each { |element|
            new_A[i][j] = element * n
            i += 1
        }
        j += 1
    }
    return new_A
end

def scale_B(n, matrix_B)
    row = matrix_B[0].length
    col = matrix_B.length
    new_B = Array.new(col){Array.new(row)}

    puts "Scaling Matrix A by n = #{n}"

    i = 0
    matrix_B.each { |array|
        j = 0
        puts array
        array.each { |element|
            new_B[i][j] = element * n
            i += 1
        }
        j += 1
    }
    return new_B
end

def det_A(matrix_A)

end


def det_B(matrix_B)

end

def transpose_A(matrix_A)
    #
end


def transpose_B(matrix_B)
    #
end


def inverse_A(matrix_A)
    # check if matrix is square
    if is_square(matrix_A)
        puts " "
    else
        puts "\n"
        puts " "
    end
end


def inverse_B(matrix_B)
    if is_square(matrix_B)
        puts " "
    else
        puts " \n"
        puts " "
    end
end


def power_A(matrix_A)

end

def power_B(matrix_B)

end

# binary operations section

# Add A and B: A + B
def add_AB(matrix_A, matrix_B)
    #
end

# Substract A from B: B - A
def sub_AB(matrix_A, matrix_B)
    #
end

# Substract B from A: A - B
def sub_BA(matrix_A, matrix_B)
    #
end

# Multiply A and B: AB
def mult_AB(matrix_A, matrix_B)
    #
end

# Multiply B and A: BA
def mult_BA(matrix_A, matrix_B)
    #
end

# copy A into B: A to B
def copy_AB(matrix_A, matrix_B)
    #
end

# copy B into A: B to A
def copy_BA(matrix_A, matrix_B)
    #
end

# Swap A and B: A to B
def swap_AB(matrix_A, matrix_B)
    #
end

# display menu
def display_menu

    puts "----------------------------------------------------------------\n" + 
        "\t\t\tMenu: Choose an option" + 
        "\n----------------------------------------------------------------"
    puts "Choice\tDescription\n\n1\tLoad Matrix A\n2\tLoad Matrix B\n" +
        "3\tSet Custom Square Matrix A\n4\tSet Custom Square Matrix B\n" + 
        "3*\tMake A and square indentity(?) matrix up to order 10x10\n" + 
        "4*\tMake B and square indentity(?) matrix up to order 10x10\n" +
        "5\tScalar n times A\n6\tScalar n times B\n" +
        "7\tDeterminant of A\n" + "8\tDeterminant of B\n" +
        "9\t\n" + "10\t\n" +
        "11\tInverse of A: A^-1\n" + "12\tInverse of B: B^-1\n" + 
        "13\t\n" +
        "14\t\n"

end

def check_continue
    valid = false
    while !valid
        print "\nDo you have more operations to perform? (Y/N): "
        cont = gets.chomp
        if cont == "Y" || cont == "y" || cont == "N" || cont == "n"
            user_input = cont
            valid = true
        else
            puts "Invalid choice. Try again"
        end
    end
    return user_input
end

def is_square(matrix)
    row = matrix[0].length
    col = matrix.length
    puts row
    puts col
    if row == col
        perfect_square = true
    else
        perfect_square = false
    end
end

# #############################################################################
# variables (may move this section)

matrix_A = Array.new()
matrix_B = Array.new()

# #############################################################################
# main script

user_input = "y"

while user_input == "y" || user_input == "Y"
    display_menu
    print "\nEnter choice selection number (1-23): "
    selection = gets.chomp

    case selection
    when '1'
        matrix_A = load_A        
        puts "\nYou have loaded Matrix A."
    when '2'
        matrix_B = load_B        
        puts "\nYou have loaded Matrix B"
    when '3'
        puts "it's a THREE, clarification required"
    when '4'
        puts "it's a FOUR, clarification required"
    when '5'
        if matrix_A.empty?
            print "\n"
        else
            print "Enter an integer to scale Matrix A: "
            user_scale_A = gets.chomp.to_i
            matrix_A = scale_A(user_scale_A, matrix_A)
        end
    when '6'
        if matrix_B.empty?
            print "\n"
        else
            print "Enter an integer to scale Matrix B: "
            user_scale_B = gets.chomp.to_i
            matrix_B = scale_B(user_scale_B, matrix_B)
        end
    when '11'
        if matrix_A.empty?
            print"\nMatrix A is EMPTY! Please Load Matrix B\n"
        else
            inverse_A(matrix_A)
        end
    when '12'
        if matrix_B.empty?
            print"\nMatrix B is EMPTY! Please Load Matrix B\n"
        else
            inverse_B(matrix_B)
        end
    else
        puts "NOT a valid choice, please try again."
    end
    user_input = check_continue
end

puts "\n\nMatrices A and B contents."
puts "Matrix A:"
print matrix_A
puts "\n"

puts "\nMatrix B:\n"
print matrix_B
puts "\n"
